services/monitor/src/DecentralizedStorageFetcher.ts

```diff
@@ -0,0 +1,113 @@
+import { FileHash } from "./util";
+import logger from "./logger";
+import { DecentralizedStorageOrigin } from "./types";
+import assert from "assert";
+import { EventEmitter } from "stream";
+import { GatewayFetcher } from "./GatewayFetcher";
+
+/**
+ * Fetcher for a certain type of Decentralized Storage (e.g. IPFS)
+ * Will try to fetch from multiple gateways.
+ */
+export default class DecentralizedStorageFetcher extends EventEmitter {
+  // TODO: Run local js-ipfs.
+  public origin: DecentralizedStorageOrigin;
+  private gatewayFetchers: GatewayFetcher[];
+  private uniqueFiles: { [key: string]: boolean } = {};
+  private uniqueFilesCounter = 0;
+  private subcriberCounter = 0;
+
+  constructor(origin: DecentralizedStorageOrigin, gateways: string[]) {
+    super();
+    this.origin = origin;
+    const fetchTimeout = parseInt(
+      process.env[`${origin.toUpperCase()}_GATEWAY_TIMEOUT`] || "30000"
+    );
+    const fetchInterval = parseInt(
+      process.env[`${origin.toUpperCase()}_GATEWAY_INTERVAL`] || "5000"
+    );
+    const fetchRetries = parseInt(
+      process.env[`${origin.toUpperCase()}_GATEWAY_RETRIES`] || "3"
+    );
+
+    this.gatewayFetchers = gateways.map(
+      (gatewayURL) =>
+        new GatewayFetcher({
+          url: gatewayURL,
+          fetchTimeout,
+          fetchInterval,
+          fetchRetries,
+        })
+    );
+  }
+
+  fetch = (fileHash: FileHash) => {
+    assert(fileHash.origin === this.origin, "Invalid origin");
+
+    return new Promise<string>((resolve, reject) => {
+      const cleanupSubscription = () => {
+        this.subcriberCounter--;
+        this.removeListener(`${fileHash.hash} fetched`, successListener);
+        this.removeListener(`${fileHash.hash} fetch failed`, failListener);
+        logger.info(
+          `Removed listener for ${fileHash.hash}. \n Unique Files counter: ${this.uniqueFilesCounter} \n Subscriber counter: ${this.subcriberCounter}`
+        );
+      };
+
+      // Only resolve on success or failure
+      const successListener = (file: string) => {
+        cleanupSubscription();
+        resolve(file);
+      };
+
+      const failListener = () => {
+        cleanupSubscription();
+        reject(`Failed to fetch ${fileHash.hash} from ${this.origin}`);
+      };
+
+      // Subscribe to fetching of the file.
+      this.on(`${fileHash.hash} fetched`, successListener);
+      // Subscribe to failure of fetching the file.
+      this.on(`${fileHash.hash} fetch failed`, failListener);
+
+      // Need the file for the first time.
+      if (!this.uniqueFiles[fileHash.hash]) {
+        this.uniqueFiles[fileHash.hash] = true;
+        this.subcriberCounter++;
+        this.uniqueFilesCounter++;
+
+        // TODO: Handle multiple gateways.
+        const gatewayFetcher = this.gatewayFetchers[0];
```

Comment:
If I understood correctly we can pass multiple IPFS gateways, but right now we are using only the first one, right?

I think that it makes sense to have at least one IPFS fallback in the first release, we know IPFS gateways are not stable.

We could implement it so that it calls different gateways concurrently and the first one that replies makes the other stop. If no one replies we can safely say that the file is not available on IPFS. For the future, this also makes sense when using a local ipfs node, we could also optimize by, for example, adding as peers all the nodes that successfully respond while the local node cannot reply.